src/atomics/relaxed.md

```diff
@@ -0,0 +1,452 @@
+let x = AtomicU32::new(0);
+let y = AtomicU32::new(0);
+
+// Thread 1:
+let r1 = y.load(atomic::Ordering::Relaxed);
+x.store(r1, atomic::Ordering::Relaxed);
+
+// Thread 2:
+let r2 = x.load(atomic::Ordering::Relaxed);
+y.store(r2, atomic::Ordering::Relaxed);
+```
+
+When starting to draw a diagram for a possible execution of this program, we
+have to first lay out the basic facts that we know:
+- `x` and `y` both start out as zero
+- Thread 1 performs a load of `y` followed by a store of `x`
+- Thread 2 performs a load of `x` followed by a store of `y`
+- Each of `x` and `y` take on exactly two values in their lifetime
+
+Then we can start to construct boxes:
+
+```text
+Thread 1      x       y      Thread 2
+╭───────╮   ┌───┐   ┌───┐   ╭───────╮
+│  load ├─┐ │ 0 │   │ 0 │ ┌─┤ load  │
+╰───╥───╯ │ └───┘   └───┘ │ ╰───╥───╯
+    ║     │   ?───────────┘     ║
+╭───⇓───╮ └───────────?     ╭───⇓───╮
+│ store ├───┬───┐   ┌───┬───┤ store │
+╰───────╯   │ ? │   │ ? │   ╰───────╯
+            └───┘   └───┘
+```
+
+At this point, if either of those lines were to connect to the higher box then
+the execution would be simple: that thread would forward the value to its lower
+box, which the other thread would then either read, or load the same value
+(zero) from the box above it, and we’d end up with zero in both atomics. But
+what if they were to connect downwards? Then we’d end up with an execution that
+looks like this:
+
+```text
+Thread 1      x       y      Thread 2
+╭───────╮   ┌───┐   ┌───┐   ╭───────╮
+│  load ├─┐ │ 0 │   │ 0 │ ┌─┤ load  │
+╰───╥───╯ │ └───┘   └───┘ │ ╰───╥───╯
+    ║     │   ┌───────────┘     ║
+╭───⇓───╮ └───┼───────┐     ╭───⇓───╮
+│ store ├───┬─┴─┐   ┌─┴─┬───┤ store │
+╰───────╯   │ ? │   │ ? │   ╰───────╯
+            └───┘   └───┘
+```
+
+But hang on — it’s not fully resolved yet, we still haven’t put in a value in
+those lower question marks. So what value should it be? Well, the second value
+of `x` is just copied from from the second value of `y`, so we just have to find
+the value of that — but the second value of `y` is itself copied from the second
+value of `x`! This means that we can actually put any value we like in that box,
+including `0` or `42`, and the logic will check out perfectly fine — meaning if
+this program were to execute in this fashion, it would end up reading a value
+produced out of thin air!
+
+Now, if we were to strictly follow the rules we’ve laid out thus far, then this
+would be totally valid thing to happen. But luckily, the authors of the C++
+specification have recognized this as a problem, and as such refined the
+semantics of `Relaxed` to implement a thorough, logically sound, mathematically
+proven formal model that prevents it, that’s just too complex and technical to
+explain here—
+
+> No “out-of-thin-air” values can be computed that circularly depend on their
+> own computations.
+
+Just kidding. Turns out, it’s a *really* difficult problem to solve, and to my
```

Comment:
It would be nice to link to https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3710.html from "difficult"